src/os/bluestore/BlueStore.cc

```diff
@@ -13064,6 +13048,22 @@ void BlueStore::_do_write_small(
   uint32_t alloc_len = min_alloc_size;
   auto offset0 = p2align<uint64_t>(offset, alloc_len);
 
+  // On zoned devices, the first goal is to support non-overwrite workloads,
+  // such as RGW, with large, aligned objects.  Therefore, for user writes
+  // _do_write_small should not trigger.  OSDs, however, write and update a tiny
+  // amount of metadata, such as OSD maps, to disk.  For those cases, we
+  // temporarily just pad them to min_alloc_size and write them to a new place
+  // on every update.
+  if (bdev->is_smr()) {
+    BlobRef b = c->new_blob();
+    uint64_t b_off = p2phase<uint64_t>(offset, alloc_len);
+    uint64_t b_off0 = b_off;
+    _pad_zeros(&bl, &b_off0, min_alloc_size);
+    o->extent_map.punch_hole(c, offset, length, &wctx->old_extents);
```

Comment:
you should call o->extent_map.fault_range(db, offset, length) prior to punching a hole to make sure proper subset of extent map is loaded.

Reply:
Done.  Moved the block below the existing extent_map.fault_range() call.